ompi/mca/io/romio321/romio/adio/ad_oceanfs/ad_oceanfs_aggrs.c

```diff
@@ -0,0 +1,552 @@
+    }
+
+    /* determine the "file domain (FD)" of each process, i.e., the portion of
+       the file that will be "owned" by each process */
+    ADIO_Offset oceanfs_ub = (max_end_offset + blksize - 1) / blksize * blksize - 1;
+    ADIO_Offset oceanfs_lb = min_st_offset / blksize * blksize;
+    ADIO_Offset oceanfs_ub_rdoff = (max_end_offset + blksize - 1) / blksize * blksize - 1 - max_end_offset;
+    ADIO_Offset oceanfs_lb_rdoff = min_st_offset - min_st_offset / blksize * blksize;
+    ADIO_Offset fd_oceanfs_range = oceanfs_ub - oceanfs_lb + 1;
+
+    int naggs = nprocs_for_coll;
+
+    /* Tweak the file domains so that no fd is smaller than a threshold.  We
+     * have to strike a balance between efficency and parallelism: somewhere
+     * between 10k processes sending 32-byte requests and one process sending a
+     * 320k request is a (system-dependent) sweet spot
+
+     This is from the common code - the new min_fd_size parm that we didn't implement.
+     (And common code uses a different declaration of fd_size so beware)
+
+     if (fd_size < min_fd_size)
+     fd_size = min_fd_size;
+     */
+    fd_size = (ADIO_Offset *)ADIOI_Malloc(nprocs_for_coll * sizeof(ADIO_Offset));
+    *fd_start_ptr = (ADIO_Offset *)ADIOI_Malloc(nprocs_for_coll * sizeof(ADIO_Offset));
+    *fd_end_ptr = (ADIO_Offset *)ADIOI_Malloc(nprocs_for_coll * sizeof(ADIO_Offset));
+    fd_start = *fd_start_ptr;
+    fd_end = *fd_end_ptr;
+
+    /* each process will have a file domain of some number of OCEANFS blocks, but
+     * the division of blocks is not likely to be even.  Some file domains will
+     * be "large" and others "small"
+     *
+     * Example: consider  17 blocks distributed over 3 aggregators.
+     * nb_cn_small = 17/3 = 5
+     * naggs_large = 17 - 3*(17/3) = 17 - 15  = 2
+     * naggs_small = 3 - 2 = 1
+     *
+     * and you end up with file domains of {5-blocks, 6-blocks, 6-blocks}
+     *
+     * what about (relatively) small files?  say, a file of 1000 blocks
+     * distributed over 2064 aggregators:
+     * nb_cn_small = 1000/2064 = 0
+     * naggs_large = 1000 - 2064*(1000/2064) = 1000
+     * naggs_small = 2064 - 1000 = 1064
+     * and you end up with domains of {0, 0, 0, ... 1, 1, 1 ...}
+     *
+     * it might be a good idea instead of having all the zeros up front, to
+     * "mix" those zeros into the fd_size array.  that way, no pset/bridge-set
+     * is left with zero work.  In fact, even if the small file domains aren't
+     * zero, it's probably still a good idea to mix the "small" file domains
+     * across the fd_size array to keep the io nodes in balance */
+    ADIO_Offset n_oceanfs_blk = fd_oceanfs_range / blksize;
+    ADIO_Offset nb_cn_small = n_oceanfs_blk / naggs;
+    ADIO_Offset naggs_large = n_oceanfs_blk - naggs * (n_oceanfs_blk / naggs);
+
+    for (i = 0; i < naggs; i++) {
+        if (i < naggs_large) {
+            fd_size[i] = (nb_cn_small + 1) * blksize;
```

Comment:
Since (nb_cn_small + 1) * blksize; and (nb_cn_small * blksize) are constant values in the loop, they can be pre-calculated and assigned into 2 variables that will be used inside the loop.

Reply:
(nb_cn_small + 1) * blksize; and (nb_cn_small * blksize) are constant values，I have defined variables to save the results to avoid repeated calculations in the loop;